other/pascal_triangle.py

```diff
@@ -90,6 +111,66 @@ def calculate_current_element(
     current_row[current_col_idx] = above_to_left_elt + above_to_right_elt
 
 
+def generate_pascal_triangle_optimized(num_rows: int) -> list[list[int]]:
+    """
+    This function returns a matrix representing the corresponding pascal's triangle
+    according to the given input of number of rows of Pascal's triangle to be generated.
+    It reduces the operations done to generate a row by half
+    by eliminating redundant calculations.
+
+    :param num_rows: Integer specifying the number of rows in the Pascal's triangle
+    :return: 2-D List (matrix) representing the Pascal's triangle
+
+    Return the Pascal's triangle of given rows
+    >>> generate_pascal_triangle_optimized(3)
+    [[1], [1, 1], [1, 2, 1]]
+    >>> generate_pascal_triangle_optimized(1)
+    [[1]]
+    >>> generate_pascal_triangle_optimized(0)
+    []
+    >>> generate_pascal_triangle_optimized(-5)
+    Traceback (most recent call last):
+        ...
+    ValueError: The input value of 'num_rows' should be greater than or equal to 0
+    >>> generate_pascal_triangle_optimized(7.89)
+    Traceback (most recent call last):
+        ...
+    TypeError: The input value of 'num_rows' should be 'int'
+    """
+
+    if not isinstance(num_rows, int):
+        raise TypeError("The input value of 'num_rows' should be 'int'")
+
+    if num_rows == 0:
+        return []
+    elif num_rows < 0:
+        raise ValueError(
+            "The input value of 'num_rows' should be greater than or equal to 0"
+        )
+
+    result: list[list[int]] = [[1]]
+
+    for row_index in range(1, num_rows):
+        temp_row = [0] + result[-1] + [0]
+        row_length = row_index + 1
+
+        # It calculates the number of distinct elements in a row
+        distinct_elements = (
+            row_length // 2 if row_length % 2 == 0 else row_length // 2 + 1
+        )
```

Comment:
Leverage the [Python builtin functions](https://docs.python.org/3/library/functions.html):
```suggestion
        distinct_elements = sum(divmod(row_length, 2))
```